### ISSUE TITLE
Helm install/upgrade keeps on waiting for Job to complete 

### ISSUE THREAD

**Issue by shivakumargowda12** [helm/helm#5891]
Have created a Job with the following template definition which is part of an existing chart.
I want this Job to run a bootstrap python script on fresh install or upgrade.

**Problem:** helm upgrade keeps on waiting for the Job to complete, inturn python script has to  complete (Its taking 10min to complete).

apiVersion: batch/v1
kind: Job
metadata:
  name: bootstrap
  annotations:
    "helm.sh/hook": "post-install,post-upgrade"
    "helm.sh/hook-delete-policy": "before-hook-creation"
spec:
  template:
    metadata:
      labels:
        app: bootstrap
    spec:
      restartPolicy: OnFailure
      containers:
        - name: bootstrap
          image: custom-bootstrap-image
          imagePullPolicy: Always

**Comment by bacongobbler** [comment]
How are you installing the job? Can you provide more information on how you install the chart?

**Comment by shivakumargowda12** [comment]
helm upgrade bt-namespace-bootstrap helm/bootstrap  --namespace bt-namespace
**Release name**: bt-namespace-bootstrap
**Chart name**: bootstrap
**Namespace**: bt-namespace
"custom-bootstrap-image" image has a python script which takes 10 minutes to complete it.
So, the above command waiting that much time to return success. It waits till the Job state is complete.

**Comment by ssalaues** [comment]
@bacongobbler Seeing the same behavior with a nearly identical looking Job definition (using post-install Helm hooks). A regular helm install without the `--wait` flag will still wait for the Job completion before the command completes. 
```
$ helm install --name my-chart . -f options.yaml
```
Helm Version
```
Server: &version.Version{SemVer:"v2.13.1", GitCommit:"618447cbf203d147601b4b9bd7f8c37a5d39fbb4", GitTreeState:"clean"}
```
Kubernetes Version
```
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.3", GitCommit:"5e53fd6bc17c0dec8434817e69b04a25d8ae0ff0", GitTreeState:"clean", BuildDate:"2019-06-06T01:44:30Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"darwin/amd64"}
```

**Comment by ikulis** [comment]
I have same problem in never version
```
$ helm version
Client: &version.Version{SemVer:"v2.14.3", GitCommit:"0e7f3b6637f7af8fcfddb3d2941fcc7cbebb0085", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.14.3", GitCommit:"0e7f3b6637f7af8fcfddb3d2941fcc7cbebb0085", GitTreeState:"clean
```

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by ssalaues** [comment]
> However, it would be nice to have the ability to tell helm NOT to wait for a job to run to completion. I suppose a workaround would be to have a sort of wrapper job that simply created the actual job and exited, but first class support would be better.

Yeah exactly, while it is well documented, it would be nice to have  some kind of option to not block/wait. I ended up switching my `post-install` job to simply a `post-install` pod that runs to completion as my work around.

**Comment by feikesteenbergen** [comment]
> Yeah exactly, while it is well documented, it would be nice to have some kind of option to not block/wait. I ended up switching my `post-install` job to simply a `post-install` pod that runs to completion as my work around.

I don't expect that to work for a Pod, if I read the documentation, @ssalaues does it actually work?

>  If the resource is a Job or Pod kind, Helm will wait until it successfully runs to completion
https://helm.sh/docs/topics/charts_hooks/#hooks-and-the-release-lifecycle

**Comment by ssalaues** [comment]
@feikesteenbergen It's been a few months since I last tested this but yes it was working as i described for the pod. If I recall correctly, the hooks waits for the deploy status of the job which is dependent on the job completion whereas just a pod reports its status immediately after scheduling/startup allowing for that specific work around.

Here's the chart and specific code however I eventually up changing back to a job because the requirements for the process changed:
https://github.com/scality/Zenko/commit/3f5182913a52ea7b441190a70652ce496c22d9d3#diff-7f002560babf32577444306436437bec

**Comment by smaslennikov** [comment]
I've got a long-running Job I must deploy. Moving it to a Pod isn't a great option as I need to preserve parallelism. An option for not waiting for a Job to finish would be awesome to have.

**Comment by TBBle** [comment]
Due to this issue, there is no good way to include a `Job` in a Helm chart that must not block install, but for which the Helm install status does not depend on the Job outcome. Per [my comment elsewhere](https://github.com/helm/helm/issues/7082#issuecomment-698416747), you cannot cleanly include a `Job` in a Helm chart as a regular resource, as many of the fields that might otherwise be modified (such as the relevant PodSpec) are immutable, so you can't include even Chart-version metadata in the resulting pod, which is really annoying.

Because a `Job` in a hook is always blocking, turning such a `Job` into a hook encounters this issue, and if it's a long-running `Job`, will probably fail the timeout.

As noted above, not all `Job`s can be turned into `Pod`s to (by stealth) remove the blocking behaviour.

Simple solution would be to have `post-*` hooks support another annotation, similar to `helm.sh/hook-delete-policy`. Perhaps `helm.sh/hook-execution-policy: non-blocking` would indicate that Helm should not depend on the outcome of the hook for its install status, which means it does not need to block on the hook.

I'm not sure such a policy would make sense for the `pre-*` hooks or the `test` hook, since running them in "pre" suggests we need an outcome from the hook before the chart itself is installed.

**Comment by TBBle** [comment]
Well, the problem hasn't gone away by itself, as far as I know.

**Comment by greggbjensen** [comment]
This is still an issue.  We have run into it as well.  We would like to start a job, but not wait for it.  The job has a process that calls back a webhook when done.

**Comment by rbren** [comment]
Can confirm this is still an issue.

My interim solution is to create a CronJob instead of a Job, with `suspended: true`. Then we have another job to `kubectl create job --from=cj/foo`.

**Comment by nicksc423** [comment]
Still an issue...

**Comment by thesuperzapper** [comment]
I resolved this issue for the [airflow-helm](https://github.com/airflow-helm/charts/tree/main/charts/airflow) chart by creating a value called `helmWait`, which removes the `helm.sh/hook-...` annotations if enabled, but only for the hooks which actually need to run BEFORE other Pods are in a Ready state:
https://github.com/airflow-helm/charts/blob/airflow-8.0.9/charts/airflow/templates/jobs/job-upgrade-db.yaml#L5-L13

Also, see https://github.com/helm/chart-testing/issues/202 for another example of an issue this behaviour is causing.

**Comment by rbren** [comment]
Ping - I think this is a really worthwhile feature

**Comment by chris-parker-za** [comment]
Agreed with @rbren

**Comment by h0jeZvgoxFepBQ2C** [comment]
Yeah we also need this feature. We would like to start a rspec automated test run after a deployment has happened.. right now its blocking and time-outing at the end.

**Comment by h0jeZvgoxFepBQ2C** [comment]
still open

**Comment by joejulian** [comment]
I've been scanning through old `question/support` issues and noticed some (this one included) suggest a feature in the comments. I've also noticed that the issues that do that, get very little notice once those `question/support` issues have aged off the 1st page.

Might I suggest that this issue be allowed to close and a new issue be created by someone in this thread that has a sufficient understanding of the problem and solution to propose a feature request (bonus points if you can also help contribute the feature)?

**Comment by rbren** [comment]
Done! https://github.com/helm/helm/issues/10555
